File: tests/benchmark/corpus/python/safe/safe_canonicalise_rooted_startswith.py
# py-safe-canonicalise-rooted: os.path.realpath + .startswith with a
# non-literal root variable (an opaque prefix-lock).  Combined with
# realpath's dotdot=No proof, is_path_traversal_safe should suppress the
# FILE_IO sink even though the canonicalised path is absolute.
import os
from flask import Flask, request

UPLOAD_ROOT = "/srv/uploads"
app = Flask(__name__)


@app.route("/file")
def file():
    name = request.args.get("name", "")
    target = os.path.realpath(os.path.join(UPLOAD_ROOT, name))
    if not target.startswith(UPLOAD_ROOT):
        return "forbidden", 403
    with open(target) as f:
        return f.read()

File: tests/benchmark/corpus/javascript/safe/safe_canonicalise_rooted_startsWith.js
// js-safe-canonicalise-rooted: path.resolve + .startsWith with a
// non-literal root variable (an opaque prefix-lock).  Combined with
// path.resolve's dotdot=No proof, is_path_traversal_safe should suppress
// the FILE_IO sink even though the canonicalised path is absolute.
const fs = require("fs");
const path = require("path");

const UPLOAD_ROOT = path.resolve("/srv/uploads");

function serveFile(req, res) {
  const name = req.query.name;
  const target = path.resolve(path.join(UPLOAD_ROOT, name));
  if (!target.startsWith(UPLOAD_ROOT)) {
    res.status(403).end();
    return;
  }
  fs.readFile(target, (err, data) => res.send(data));
}

module.exports = { serveFile };
